submitit/snapshot.py

```diff
@@ -0,0 +1,38 @@
+import fnmatch
+import os
+import shutil
+from subprocess import check_output
+
+
+def run_cmd(str_args, **kwargs):
+    return check_output(str_args, shell=True, **kwargs).decode("utf-8").strip()
+
+
+class SnapshotManager:
+    def __init__(
+        self, snapshot_dir: str, with_submodules: bool = False, exclude=None,
+    ):
+        self.snapshot_dir = snapshot_dir
+        self.original_dir = os.getcwd()
+        self.with_submodules = with_submodules
+        self.exclude = exclude or []
+
+    def __enter__(self):
+        self.original_dir = os.getcwd()
+        os.makedirs(self.snapshot_dir, exist_ok=True)
+        # Get the repository root
+        root_dir = run_cmd("git rev-parse --show-toplevel")
+        sub = "--recurse-submodules" if self.with_submodules else "-s"
+        # https://stackoverflow.com/a/51689219/4876946
+        files = run_cmd(f"git ls-files {sub} | grep -v ^16 | cut -f2-", cwd=root_dir).split("\n")
+        # Copy all the files into the snapshot directory
+        for file in files:
+            if any([fnmatch.fnmatch(file, ex) for ex in self.exclude]):
+                continue
+            dest_file = os.path.join(self.snapshot_dir, file)
+            os.makedirs(os.path.dirname(dest_file), exist_ok=True)
+            shutil.copy(os.path.join(root_dir, file), dest_file)
```

Comment:
Rsync of a bunch of files together is likely faster

Reply:
If we use Rsync then I think Rsync should appear in the name just to make it clear on the user.

Reply:
Also rsync can respect .gitignore which we probably want to do to avoid copying models/checkpoints in the root dir: `rsync --exclude='.git/' --filter=':- .gitignore'`

Reply:
Thanks @stephenroller for the suggestion!  I'm getting a list of all the checked in files using `git ls-files` and stashing them in a temporary file that gets passed to rsync now.  I think this should address the above comments (fast and avoid copying models/checkpoints etc).

Reply:
> If we use Rsync then I think Rsync should appear in the name just to make it clear on the user.

I would also tend to check that Rsync is installed in the init, and provide a nice message as error explaining that it is needed and how to install it? Would be a nice user experience